plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java

```diff
@@ -0,0 +1,152 @@
+                        .add(GET_DATABASE)
+                        .addCopies(GET_TABLE, 2)
+                        .build());
+    }
+
+    @Test
+    public void testCreateTableAsSelect()
+    {
+        assertMetastoreInvocations("CREATE TABLE test_ctas AS SELECT 1 AS age",
+                ImmutableMultiset.builder()
+                        .add(GET_DATABASE)
+                        .add(CREATE_TABLE)
+                        .addCopies(GET_TABLE, 2)
+                        .build());
+    }
+
+    @Test
+    public void testSelect()
+    {
+        assertMetastoreInvocations("CREATE TABLE test_select_from (id VARCHAR, age INT)",
+                ImmutableMultiset.builder()
+                        .add(GET_DATABASE)
+                        .add(CREATE_TABLE)
+                        .addCopies(GET_TABLE, 2)
+                        .build());
+
+        assertMetastoreInvocations("SELECT * FROM test_select_from",
+                ImmutableMultiset.builder()
+                        .addCopies(GET_TABLE, 12)
+                        .build());
+    }
+
+    @Test
+    public void testSelectWithFilter()
+    {
+        assertMetastoreInvocations("CREATE TABLE test_select_from_where AS SELECT 2 as age",
+                ImmutableMultiset.builder()
+                        .add(GET_DATABASE)
+                        .add(CREATE_TABLE)
+                        .addCopies(GET_TABLE, 2)
+                        .build());
+
+        assertMetastoreInvocations("SELECT * FROM test_select_from_where WHERE age = 2",
+                ImmutableMultiset.builder()
+                        .addCopies(GET_TABLE, 12)
+                        .build());
+    }
+
+    @Test
+    public void testJoin()
+    {
+        assertMetastoreInvocations("CREATE TABLE test_join_t1 AS SELECT 2 as age, 'id1' AS id",
+                ImmutableMultiset.builder()
+                        .add(GET_DATABASE)
+                        .add(CREATE_TABLE)
+                        .addCopies(GET_TABLE, 2)
+                        .build());
+        assertMetastoreInvocations("CREATE TABLE test_join_t2 AS SELECT 'name1' as name, 'id1' AS id",
+                ImmutableMultiset.builder()
+                        .add(GET_DATABASE)
+                        .add(CREATE_TABLE)
+                        .addCopies(GET_TABLE, 2)
+                        .build());
+
+        assertMetastoreInvocations("SELECT name, age FROM test_join_t1 JOIN test_join_t2 ON test_join_t2.id = test_join_t1.id",
+                ImmutableMultiset.builder()
+                        .addCopies(GET_TABLE, 24)
+                        .build());
+    }
+
+    private void assertMetastoreInvocations(String query, Multiset expectedInvocations)
+    {
+        try {
+            getQueryRunner().execute(query);
+            assertThat(metastore.getMethodInvocations()).containsExactlyInAnyOrderElementsOf(expectedInvocations);
+        }
+        finally {
+            metastore.resetCounters();
```

Comment:
Reset counters first (before running a query)

- then you can run preparations without wrapping them in `assertMetastoreInvocations`
  - actually you can do this today, just next `assertMetastoreInvocations` invocation would return incorrect results
- no need for `finally`

Reply:
sure, good idea